### ISSUE TITLE
NumPy Ufuncs on scalar values don't work in CUDA

### ISSUE THREAD

**Issue by gmarkall** [numba/numba#7112]
Reported in [this Discourse post](https://numba.discourse.group/t/cant-use-basic-numpy-functions-with-cuda-like-zeros-or-empty/628/4).

The [CUDA NumPy support documentation](https://numba.readthedocs.io/en/latest/cuda/cudapysupported.html#numpy-support) states that:

> scalar ufuncs that have equivalents in the math module; i.e. np.sin(x[0]), where x is a 1D array.

are supported. However, executing:

```python
import numpy as np
from numba import cuda
import math

test_in2 = cuda.to_device(np.array([0, math.pi/3, math.pi]))
test_out2 = cuda.to_device(np.array([0,0,0]))

@cuda.jit
def test_math_sin_numpy(in_array, out_array):
    i = cuda.grid(1)
    if i < in_array.size:
        if np.sin(in_array[i]) < .5:
            out_array[i] = 100

test_math_sin_numpy.forall(test_in2.size)(test_in2, test_out2)
print(test_out2.copy_to_host())
```

results in an exception:

```
numba.core.errors.TypingError: Failed in cuda mode pipeline (step: nopython frontend)
Use of unsupported NumPy function 'numpy.sin' or unsupported use of the function.

File "repro.py", line 12:
def test_math_sin_numpy(in_array, out_array):
    <source elided>
    if i < in_array.size:
        if np.sin(in_array[i]) < .5:
        ^

During: typing of get attribute at /home/gmarkall/numbadev/issues/discourse-628/repro.py (12)

File "repro.py", line 12:
def test_math_sin_numpy(in_array, out_array):
    <source elided>
    if i < in_array.size:
        if np.sin(in_array[i]) < .5:
```

This has never worked since the relevant documentation was added in #2722. However, it should not be too difficult to fix - the typing and lowering for these NumPy ufuncs can share the implementations from the typing and lowering for the `math` module.

**Comment by yaredev0** [comment]
Would implementing this indirectly also address [#4333](https://github.com/numba/numba/issues/4333)

**Comment by gmarkall** [comment]
> Would implementing this indirectly also address [#4333](https://github.com/numba/numba/issues/4333)?

I don't think so, I think they're distinct... Are you interested in looking at either of these issues?

**Comment by yaredev0** [comment]
> I don't think so, I think they're distinct... 

Hm. I only brought it up because [numpy.isnan](https://numpy.org/doc/stable/reference/generated/numpy.isnan.html) and [numpy.isnat](https://numpy.org/doc/stable/reference/generated/numpy.isnat.html) are numpy ufuncs

> Are you interested in looking at either of these issues?

I would be interested in contributing, I actually first encountered this issue during an initial dive into the `good first issue` tags to see if I could figure it out.

I don't have experience with LLVM or the numba codebase though. Would https://numba.pydata.org/numba-doc/dev/extending/index.html be what I should be reading to see how to implement these?

**Comment by gmarkall** [comment]
> I don't have experience with LLVM or the numba codebase though. Would https://numba.pydata.org/numba-doc/dev/extending/index.html be what I should be reading to see how to implement these?

That's a useful reference, but I'd be inclined to suggest looking at a couple of others first:

* [The Life of a Numba Kernel: A Compilation Pipeline Taking User Defined Functions in Python to CUDA Kernels](https://medium.com/rapids-ai/the-life-of-a-numba-kernel-a-compilation-pipeline-taking-user-defined-functions-in-python-to-cuda-71cc39b77625) - this gives an overview of the whole pipeline, with more detail in [the accompanying notebook].(https://github.com/gmarkall/life-of-a-numba-kernel/blob/master/Life%20of%20a%20Numba%20Kernel%20-%20with-%20output.ipynb)
* [Section 4](https://github.com/numba/nvidia-cuda-tutorial/blob/main/session-4/session-4-with-notes.pdf) of [Numba for CUDA Programmers](https://github.com/numba/nvidia-cuda-tutorial), which explains how to implement support for types and functions in the CUDA target.
* It's probably also worth looking at a couple of PRs that add support to the CUDA target, to see what additional typing and lowering looks like, e.g.:
   - [`log2` and `remainder` additions](https://github.com/numba/numba/pull/6358)
   - [`isnan` and `isinf` on integers](https://github.com/numba/numba/pull/5761)

If you start looking through these and working on adddition of the typing and lowering either for these ufuncs on scalars, or `isnat`, and you have questions, please do drop into the [Gitter](https://gitter.im/numba/numba) and ask questions, or to start a thread on [Discourse](https://numba.discourse.group/) - feel free to ping me on either of those platforms as well.

**Comment by yaredev0** [comment]
Sorry for the late reply, I'll take a look over the weekend and try to see if I can figure out how to tackle this issue.

**Comment by yaredev0** [comment]
I gave the docs and the PRs a quick readthrough, but I am not fully sure where to start. 

1. From what I can figure, there's no typing or lowering at all for the numpy ufuncs. I am assuming that the `math` module' is being done through `cuda.cudamath.py` and `cuda.mathimpl.py`. When you say "the typing and lowering for these NumPy ufuncs can share the implementations from the typing and lowering for the math module", do you mean that the numpy code can be dropped into these files (or placed into another new one?)

2. There is a [ufunc like object](https://numba.pydata.org/numba-doc/dev/user/vectorize.html) specifically for dealing with cuda. Would this be of any use or relevance to this?
